tools/src/main/java/org/apache/kafka/trogdor/workload/SustainedConnectionWorker.java

```diff
@@ -0,0 +1,563 @@
+                    this.partitions = this.producer.partitionsFor(this.topicName).stream()
+                             .map(partitionInfo -> new TopicPartition(partitionInfo.topic(), partitionInfo.partition()))
+                             .collect(Collectors.toList());
+                    Collections.shuffle(this.partitions);
+                }
+
+                // Create a new iterator over the partitions if the current one doesn't exist or is exhausted.
+                if (this.partitionsIterator == null || !this.partitionsIterator.hasNext()) {
+                    this.partitionsIterator = this.partitions.iterator();
+                }
+
+                // Produce a single record and send it synchronously.
+                TopicPartition partition = this.partitionsIterator.next();
+                ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
+                        partition.topic(), partition.partition(), keys.next(), values.next());
+                producer.send(record).get();
+
+            } catch (Throwable e) {
+                // Set the producer to be recreated on the next cycle.
+                this.closeQuietly();
+
+                // Housekeeping to track the number of opened connections and failed connection attempts.
+                SustainedConnectionWorker.this.totalProducerConnections.decrementAndGet();
+                SustainedConnectionWorker.this.totalProducerFailedConnections.incrementAndGet();
+                SustainedConnectionWorker.log.error("Error while refreshing sustained KafkaProducer connection", e.getMessage());
+
+            }
+
+            // Schedule this again and set to not in use.
+            this.nextUpdate = System.currentTimeMillis() + this.refreshRate;
+            this.inUse = false;
+        }
+
+        @Override
+        protected void closeQuietly() {
+            Utils.closeQuietly(this.producer, "KafkaProducer");
+            this.producer = null;
+            this.partitions = null;
+            this.partitionsIterator = null;
+        }
+    }
+
+    private class ConsumerSustainedConnection extends ClaimableConnection {
+
+        // These variables are used to maintain the connection itself.
+        private String topicName;
+        private KafkaConsumer<byte[], byte[]> consumer;
+        private String consumerGroup;
+        private TopicPartition activePartition;
+        private Random rand;
+
+        // This variable is used to maintain the scheduling of the connection keep-alive.
+        private long refreshRate;
+
+        // This variable is used to maintain the connection properties.
+        private final Properties props;
+
+        ConsumerSustainedConnection() {
+            // These variables are used to maintain the connection itself.
+            this.topicName = SustainedConnectionWorker.this.spec.topicName();
+            this.consumer = null;
+            this.consumerGroup = UUID.randomUUID().toString();
```

Comment:
Because we're not using the `subscribe()` functionality, we don't need to configure a consumer group. It will use a null group by default (https://cwiki.apache.org/confluence/display/KAFKA/KIP-289%3A+Improve+the+default+group+id+behavior+in+KafkaConsumer)

Reply:
Done.